### PR TITLE
New validation for REST API spec location before SDK release

### PR DESCRIPTION
**Pull request by raych1**
We only support the SDK release for the spec checked into 'main' branch of Azure/azure-rest-api-specs repository.

- Added common script for validation on `tsp-location.yaml` (typespec case) and validation on `autorest.md` (swagger case)
- Added common pipeline

@weshaggard can you review?

### CODE REVIEW

File: eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml:14]:
```suggestion
        -ServiceDirectory '${{ parameters.ServiceDirectory }}'
```

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:56]: Cloning down the specs repo in this check seems pretty heavy handed. Is there a github api we can call to check if a it is part of main? Or perhaps we could use the sparse clone from the generation command? We could even potentially add this as a check in the generate command.

Comment (weshaggard) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:56]: Perhaps you could use the search api as outlined in https://stackoverflow.com/questions/30376787/github-api-check-if-a-branch-or-repository-contains-a-commit

Comment (weshaggard) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:56]: Example:
`https://api.github.com/search/commits?q=repo:Azure/azure-sdk-tools+hash:dff04c39ef72f91c27cb10f164b10b6bf958edc9`

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: Right, I also thought that git clone is a little bit heavier. I looked at this search api previously but hesitated to use it concerning the paging behavior. I think we will not hit the paging case for current use case.

File: eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml:23]: You can remove this given fals is the default.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:54]: Is the PAT needed here? Are we worried about rate limiting?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:87]: Do we need to worry about multiple yaml secions?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:110]: How do we not show these warnings for any project that isn't generated? We need to be careful to not start causing a bunch of noise in the builds.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:87]: Yes, we take the first element of it if having multiple sections. autorest.md doesn't have a title for this yaml section.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:54]: I prefer to provide the PAT given the rate limit is three times difference. 10 calls vs 30 calls.  I would like to avoid failing the pipeline by this rate limit issue.

Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:54]: Why do we have a hard-coded commit hash here?

Comment (konrad-jamrozik) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:54]: If we keep it, I would appreciate adding in a comment an URL pointing to this commit in the Web UI.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: Iterate all yaml sections now.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: Removed.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (lirenhe) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:167]: Is the handling of JS the same with Java?
I found JS is using _meta.json file  [sdk/agrifood/arm-agrifood/_meta.json](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/agrifood/arm-agrifood/_meta.json)

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:167]: Thanks @lirenhe for the reminder. The config format appears to be different from `readme.md` and `autorest.md`.

@qiaozha , can you confirm if `_meta.json` is the configuration file of the spec for JS management plane SDK?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:26]: You should be able to make these into one regex instead of needing to split this into multiple conditions. 
```
^https://(raw.githubusercontent.com|github.com)/(?<repo>[^/]*/azure-rest-api-specs)/(blob|tree)?/(?<commit>[^\/]+(\/[^\/]+)*|[0-9a-f]{40})/(?<path>specification/.*)
```

**Review by weshaggard — COMMENTED** [review]
This looks reasonable to me I'll leave it to Renhe to sign-off as he will have more knowledge about the content of these files.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (qiaozha) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:167]: we store the commit id for swagger in `_meta.json` , which is for regeneration for codegen change only, if it's for pipeline, I think we should always use the lastest commit id ?

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:167]: @qiaozha 
Do you mean `_meta.json` is only used for re-generation when codegen changes?  
I would like to know is there any config of spec location used in JS SDK release phase? That would be the content of this validation to check.  
Or, does the JS generation script pull the spec from latest main branch of Azure/azure-rest-api-specs repo by default for release? If that's the case, then we can ignore this validation for JS mgmt. plane scenario.

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: @lirenhe , for Java/JS/Go mgmt. plane scenarios, they will always get the latest main branch of Azure/azure-rest-api-specs repo to release the SDK. I will ignore the validation for them.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: I thought about the combination of these two expressions but seem cannot be simplified. Anyway, I combined them with one simplified version with missing an edge case coverage. The below URL starting with `raw.githubusercontent.com` should not have `[blob|tree]` part after repository name.
```
https://raw.githubusercontent.com/Azure/azure-rest-api-specs/main/specification/purview/data-plane/Azure.Analytics.Purview.MetadataPolicies/preview/2021-07-01-preview/purviewMetadataPolicy.json
```

**Comment by raych1** [comment]
@weshaggard , I added the package version validation to stop the validation if the package is not GA version.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: That is why I made `((blob|tree)/)?` optional in my regex example.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:138]: Let's avoid using RetrievePackages here instead do something similar to what APIView does. See https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/Create-APIReview.ps1#L107C63-L107C63 it reads the metadata from the packageinfo.json file.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: I know, making it as optional cannot avoid this invalid case which should not have `blob` before `main`. However, it's ok to skip this case as this script is not the place to ensure the URL is valid.
```
https://raw.githubusercontent.com/Azure/azure-rest-api-specs/blob/main/specification/purview/data-plane/Azure.Analytics.Purview.MetadataPolicies/preview/2021-07-01-preview/purviewMetadataPolicy.json
```

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: I updated it and validated locally. Can you review? @weshaggard

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (chunyu3) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:120]: Are batchValue, requireValue, inputFileValue exclusive in a readme file? If they can exist concurrently, there will be problem.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (chunyu3) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:242]: Better to use existence of file `autorest.md` as if-condition, not the language name for extensibility in future.

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:242]: I prefer to split the cases by languages. Go language use 'autorest.md' but we will ignore it.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:120]: I haven't found a real case with having two of these configurations. Currently, I follow below sequence for the validation. 
Do you have any idea of how does the generation tool handle this? 
+@weidongxu-microsoft, any comment?

Comment (weidongxu-microsoft) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:120]: I am OK with current order.
In my understanding, we are not achieving full support for the markdown for autorest. (one may need to use the parse code from autorest to achieve full support, e.g. https://github.com/Azure/autorest/blob/main/packages/libs/configuration/src/literate-yaml/literate-yaml.test.ts)

So, as long as it works reasonably fine, I am good with it.

Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:120]: Thanks for sharing the reference of autorest code logic. I learned that autorest handles the case of `json` content in the markdown file. I would like to ignore that part because I haven't found the usage of `json` in `autorest.md` and `swagger/README.md`.

**Review by lirenhe — APPROVED** [review]

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:38]: Get the token from environment variable so that it won't be leaked in the pipeline log.

**Comment by raych1** [comment]
@weshaggard , can you also approve this PR if the change looks good?

File: eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml:20]: Why do we need to set it here as well as the env? Als it doesn't look like this parameter is listed in the template.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:39]: When do we use this parameter over the ArtifactLocation?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:214]: Do you want to exit the entire script here? What if there are multiple packages?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:281]: Please add a newline at the end of the file

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:214]: Nevermind I just looked closer at the code and there shouldn't ever be more than one package returned. I know I told you to use this same method and I do still agree with that but I may want to clean that code up at some point.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [inline eng/common/scripts/Verify-RestApiSpecLocation.ps1:179]: Is this only for local testing?

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (weshaggard) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:39]: I believe I figured out that this is for local testing.

**Review by weshaggard — APPROVED** [review]
A couple comments but overall looks good to get started using.

File: eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml
Comment (raych1) [reply eng/common/pipelines/templates/steps/verify-restapi-spec-location.yml:20]: Thanks, removed. Forgot to remove this line.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1:179]: No. This is used in release pipeline where doesn't pass in the value. Because during release job stage, it doesn't have the `packageInfo.json` generated at signed artifact folder. This script needs to generate that file before getting the package version.

File: eng/common/scripts/Verify-RestApiSpecLocation.ps1
Comment (raych1) [reply eng/common/scripts/Verify-RestApiSpecLocation.ps1]: Done.

**Pull request merged**
